File: Release/Silverlight5/SampleSource/Controls/Controls.Samples/Theming/ThemeBrowser/ThemeCatalogItem.cs
// (c) Copyright Microsoft Corporation.
// This source is subject to the Microsoft Public License (Ms-PL).
// Please see http://go.microsoft.com/fwlink/?LinkID=131993 for details.
// All other rights reserved.

using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Windows.Controls.Theming;
using System.Windows.Media;

namespace System.Windows.Controls.Samples
{
    /// <summary>
    /// Buisness Object used in the ThemeBrowserSample.
    /// </summary>
    public sealed partial class ThemeCatalogItem
    {
        /// <summary>
        ///  Gets the name of the theme.
        /// </summary>
        public string ThemeName { get; private set; }

        /// <summary>
        /// Gets the thumnail Image representing the theme.
        /// </summary>
        public Image ThumbnailImage { get; private set; }

        /// <summary>
        /// Gets a new Theme instance.
        /// </summary>
        public Func<Theme> ThemeConstructor { get; private set; }

        /// <summary>
        /// Initializes a new instance of the ThemeCatologItem class.
        /// </summary>
        public ThemeCatalogItem()
        {
        }

        /// <summary>
        /// Initializes a new instance of the ThemeCatologItem class.
        /// </summary>
        /// <param name="themeName">Name of the theme.</param>
        /// <param name="resourceName">
        /// Name of the resource defining the Thumbnail for the theme.
        /// </param>
        /// <param name="constructor">
        /// Constructs a new Theme instance.
        /// </param>
        internal ThemeCatalogItem(string themeName, string resourceName, Func<Theme> constructor)
        {
            ThemeName = themeName;
            ThumbnailImage = SharedResources.GetImage(resourceName);
            ThemeConstructor = constructor;
        }

        #region Sample Data
        /// <summary>
        /// Gets a Theme object that represents Shiny Blue.
        /// </summary>
        public static ThemeCatalogItem ShinyBlue
        {
            get
            {
                return new ThemeCatalogItem(
                    "Shiny Blue",
                    "thumbShinyBlue.png",
                    () => new ShinyBlueTheme());
            }
        }

        /// <summary>
        /// Gets a Theme object that represents Shiny Red.
        /// </summary>
        public static ThemeCatalogItem ShinyRed
        {
            get
            {
                return new ThemeCatalogItem(
                    "Shiny Red",
                    "thumbShinyRed.png",
                    () => new ShinyRedTheme());
            }
        }

        /// <summary>
        /// Gets a Theme object that represents RainierOrange Theme.
        /// </summary>
        public static ThemeCatalogItem RainierOrange
        {
            get
            {
                return new ThemeCatalogItem(
                    "Rainier Orange",
                    "thumbRainierOrange.png",
                    () => new RainierOrangeTheme());
            }
        }

        /// <summary>
        /// Gets a Theme object that represents RainierPurple Theme.
        /// </summary>
        public static ThemeCatalogItem RainierPurple
        {
            get
            {
                return new ThemeCatalogItem(
                    "Rainier Purple",
                    "thumbRainierPurple.png",
                    () => new RainierPurpleTheme());
            }
        }

        /// <summary>
        /// Gets a Theme object that represents ExpressionDark.
        /// </summary>
        public static ThemeCatalogItem ExpressionDark
        {
            get
            {
                return new ThemeCatalogItem(
                    "Expression Dark",
                    "thumbExpresssionDark.png",
                    () => new ExpressionDarkTheme());
            }
        }

        /// <summary>
        /// Gets a Theme object that represents ExpressionDark.
        /// </summary>
        public static ThemeCatalogItem ExpressionLight
        {
            get
            {
                return new ThemeCatalogItem(
                    "Expression Light",
                    "thumbExpresssionLight.png",
                    () => new ExpressionLightTheme());
            }
        }

        /// <summary>
        /// Gets a Theme object that represents Bureau Blue.
        /// </summary>
        public static ThemeCatalogItem BureauBlue
        {
            get
            {
                return new ThemeCatalogItem(
                    "Bureau Blue",
                    "BureauBlue.png",
                    () => new BureauBlueTheme());
            }
        }

        /// <summary>
        /// Gets a Theme object that represents Bureau Black.
        /// </summary>
        public static ThemeCatalogItem BureauBlack
        {
            get
            {
                return new ThemeCatalogItem(
                    "Bureau Black",
                    "BureauBlack.png",
                    () => new BureauBlackTheme());
            }
        }

        /// <summary>
        /// Gets a Theme object that represents Bureau Black.
        /// </summary>
        public static ThemeCatalogItem WhistlerBlue
        {
            get
            {
                return new ThemeCatalogItem(
                    "Whistler Blue",
                    "WhistlerBlue.png",
                    () => new WhistlerBlueTheme());
            }
        }

        /// <summary>
        /// Gets a Theme object that represents System Colors.
        /// </summary>
        public static ThemeCatalogItem SystemColors
        {
            get
            {
                return new ThemeCatalogItem(
                    "System Colors",
                    "SystemColors.png",
                    () => new SystemColorsTheme());
            }
        }

        /// <summary>
        /// Gets a Theme object that represents Twilight Blue.
        /// </summary>
        public static ThemeCatalogItem TwilightBlue
        {
            get
            {
                return new ThemeCatalogItem(
                    "Twilight Blue",
                    "TwilightBlue.png",
                    () => new TwilightBlueTheme());
            }
        }

        /// <summary>
        /// Gets a Theme object that represents Bubble Creme.
        /// </summary>
        [SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly", MessageId = "Creme", Justification = "Correct Spelling")]
        public static ThemeCatalogItem BubbleCreme
        {
            get
            {
                return new ThemeCatalogItem(
                    "Bubble Creme",
                    "BubbleCreme.png",
                    () => new BubbleCremeTheme());
            }
        }

        /// <summary>
        /// Gets a collection of theme objects.
        /// </summary>
        public static IEnumerable<ThemeCatalogItem> ThemesCatalog
        {
            get { return new ThemeCatalogItem[] { BubbleCreme, TwilightBlue, ExpressionDark, ExpressionLight, WhistlerBlue, BureauBlack, BureauBlue, ShinyBlue, ShinyRed, RainierOrange, RainierPurple, SystemColors }; }
        }
        #endregion Sample Data
    }
}

File: Release/Silverlight5/SampleSource/Controls/Controls.Samples/Theming/ThemeBrowser/ThemeCatalogItem.vb
' (c) Copyright Microsoft Corporation.
' This source is subject to the Microsoft Public License (Ms-PL).
' Please see http://go.microsoft.com/fwlink/?LinkID=131993 for details.
' All other rights reserved.

Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Diagnostics.CodeAnalysis
Imports System.Windows.Media
Imports System.Windows.Controls

''' <summary>
''' Buisness Object used in the ThemeBrowserSample.
''' </summary>
Partial Public NotInheritable Class ThemeCatalogItem
    ''' <summary>
    '''  Gets or sets the name of the theme.
    ''' </summary>
    Private privateThemeName As String
    Public Property ThemeName() As String
        Get
            Return privateThemeName
        End Get
        Set(ByVal value As String)
            privateThemeName = value
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the thumnail Image representing the theme.
    ''' </summary>
    Private privateThumbnailImage As Image
    Public Property ThumbnailImage() As Image
        Get
            Return privateThumbnailImage
        End Get
        Set(ByVal value As Image)
            privateThumbnailImage = value
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the prefered background brush for this theme.
    ''' </summary>
    Private privatePreferredBackground As Brush
    Public Property PreferredBackground() As Brush
        Get
            Return privatePreferredBackground
        End Get
        Set(ByVal value As Brush)
            privatePreferredBackground = value
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the path of the theme xaml file for this theme.
    ''' </summary>
    Private privateThemeXamlPath As String
    Public Property ThemeXamlPath() As String
        Get
            Return privateThemeXamlPath
        End Get
        Set(ByVal value As String)
            privateThemeXamlPath = value
        End Set
    End Property

    ''' <summary>
    ''' Initializes a new instance of the ThemeCatologItem class.
    ''' </summary>
    Public Sub New()
    End Sub

    ''' <summary>
    ''' Initializes a new instance of the ThemeCatologItem class.
    ''' </summary>
    ''' <param name="themeName">Name of the theme.</param>
    ''' <param name="resourceName">
    ''' Name of the resource defining the Thumbnail for the theme.
    ''' </param>
    ''' <param name="themeXamlPath">
    ''' Path to the XAML file for this theme.
    ''' </param>
    ''' <param name="preferredBackground">
    ''' Preferred Background for the theme.
    ''' </param>
    Friend Sub New(ByVal themeName As String, ByVal resourceName As String, ByVal themeXamlPath As String, ByVal preferredBackground As Brush)
        Me.ThemeName = themeName
        Me.ThumbnailImage = SharedResources.GetImage(resourceName)
        Me.ThemeXamlPath = themeXamlPath
        Me.PreferredBackground = preferredBackground
    End Sub

#Region "Sample Data"
    ''' <summary>
    ''' Gets a Theme object that represents Shiny Blue.
    ''' </summary>
    Public Shared ReadOnly Property ShinyBlue() As ThemeCatalogItem
        Get
            Dim brush As LinearGradientBrush = GetLinearBrush(Color.FromArgb(255, 132, 150, 170), Color.FromArgb(255, 43, 49, 56))
            Return New ThemeCatalogItem("Shiny Blue", "thumbShinyBlue.png", "System.Windows.Controls.Samples;component/theming/ThemeBrowser/ShinyBlue.xaml", brush)
        End Get
    End Property

    ''' <summary>
    ''' Gets a Theme object that represents Shiny Red.
    ''' </summary>
    Public Shared ReadOnly Property ShinyRed() As ThemeCatalogItem
        Get
            Dim brush As LinearGradientBrush = GetLinearBrush(Color.FromArgb(255, 209, 209, 209), Color.FromArgb(255, 174, 49, 56))
            Return New ThemeCatalogItem("Shiny Red", "thumbShinyRed.png", "System.Windows.Controls.Samples;component/Theming/ThemeBrowser/ShinyRed.xaml", brush)
        End Get
    End Property

    ''' <summary>
    ''' Gets a Theme object that represents RainierOrange Theme.
    ''' </summary>
    Public Shared ReadOnly Property RainierOrange() As ThemeCatalogItem
        Get
            Dim brush As LinearGradientBrush = GetLinearBrush(Color.FromArgb(255, 0, 0, 0), Color.FromArgb(255, 7, 15, 19))
            Return New ThemeCatalogItem("Rainier Orange", "thumbRainierOrange.png", "System.Windows.Controls.Samples;component/theming/ThemeBrowser/RainierOrange.xaml", brush)
        End Get
    End Property

    ''' <summary>
    ''' Gets a Theme object that represents RainierPurple Theme.
    ''' </summary>
    Public Shared ReadOnly Property RainierPurple() As ThemeCatalogItem
        Get
            Dim brush As LinearGradientBrush = GetLinearBrush(Color.FromArgb(255, 0, 0, 0), Color.FromArgb(255, 7, 15, 19))
            Return New ThemeCatalogItem("Rainier Purple", "thumbRainierPurple.png", "System.Windows.Controls.Samples;component/theming/ThemeBrowser/RainierPurple.xaml", brush)
        End Get
    End Property

    ''' <summary>
    ''' Gets a Theme object that represents ExpressionDark.
    ''' </summary>
    Public Shared ReadOnly Property ExpressionDark() As ThemeCatalogItem
        Get
            Dim brush As LinearGradientBrush = GetLinearBrush(Color.FromArgb(255, 23, 23, 23), Color.FromArgb(255, 10, 10, 10))
            Return New ThemeCatalogItem("Expression Dark", "thumbExpresssionDark.png", "System.Windows.Controls.Samples;component/theming/ThemeBrowser/ExpressionDark.xaml", brush)
        End Get
    End Property

    ''' <summary>
    ''' Gets a Theme object that represents ExpressionDark.
    ''' </summary>
    Public Shared ReadOnly Property ExpressionLight() As ThemeCatalogItem
        Get
            Dim brush As LinearGradientBrush = GetLinearBrush(Color.FromArgb(255, 189, 189, 189), Color.FromArgb(255, 160, 160, 160))
            Return New ThemeCatalogItem("Expression Light", "thumbExpresssionLight.png", "System.Windows.Controls.Samples;component/theming/ThemeBrowser/ExpressionLight.xaml", brush)
        End Get
    End Property

    ''' <summary>
    ''' Gets a Theme object that represents Bureau Blue.
    ''' </summary>
    Public Shared ReadOnly Property BureauBlue() As ThemeCatalogItem
        Get
            Dim brush As LinearGradientBrush = GetLinearBrush(Color.FromArgb(255, 191, 219, 255), Color.FromArgb(255, 166, 194, 229))
            Return New ThemeCatalogItem("Bureau Blue", "BureauBlue.png", "System.Windows.Controls.Samples;component/theming/ThemeBrowser/BureauBlue.xaml", brush)
        End Get
    End Property

    ''' <summary>
    ''' Gets a Theme object that represents Bureau Black.
    ''' </summary>
    Public Shared ReadOnly Property BureauBlack() As ThemeCatalogItem
        Get
            Dim brush As LinearGradientBrush = GetLinearBrush(Color.FromArgb(255, 91, 91, 91), Color.FromArgb(255, 13, 13, 13))
            Return New ThemeCatalogItem("Bureau Black", "BureauBlack.png", "System.Windows.Controls.Samples;component/theming/ThemeBrowser/BureauBlack.xaml", brush)
        End Get
    End Property

    ''' <summary>
    ''' Gets a Theme object that represents Bureau Black.
    ''' </summary>
    Public Shared ReadOnly Property WhistlerBlue() As ThemeCatalogItem
        Get
            Dim brush As LinearGradientBrush = GetLinearBrush(Color.FromArgb(255, 255, 255, 255), Color.FromArgb(255, 255, 255, 255))
            Return New ThemeCatalogItem("Whistler Blue", "WhistlerBlue.png", "System.Windows.Controls.Samples;component/theming/ThemeBrowser/WhistlerBlue.xaml", brush)
        End Get
    End Property

    ''' <summary>
    ''' Gets a Theme object that represents Twilight Blue.
    ''' </summary>
    Public Shared ReadOnly Property TwilightBlue() As ThemeCatalogItem
        Get
            Dim brush As LinearGradientBrush = GetLinearBrush(Color.FromArgb(255, 120, 183, 225), Color.FromArgb(255, 120, 183, 225))
            Return New ThemeCatalogItem("Twilight Blue", "TwilightBlue.png", "System.Windows.Controls.Samples;component/theming/ThemeBrowser/TwilightBlue.xaml", brush)
        End Get
    End Property

    ''' <summary>
    ''' Gets a Theme object that represents Bubble Creme.
    ''' </summary>
    <SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly", MessageId:="Creme", Justification:="Creme is an alternate spelling.")> _
    Public Shared ReadOnly Property BubbleCreme() As ThemeCatalogItem
        Get
            Dim brush As LinearGradientBrush = GetLinearBrush(Color.FromArgb(255, 249, 249, 230), Color.FromArgb(255, 249, 249, 230))
            Return New ThemeCatalogItem("Bubble Creme", "BubbleCreme.png", "System.Windows.Controls.Samples;component/theming/ThemeBrowser/BubbleCreme.xaml", brush)
        End Get
    End Property

    ''' <summary>
    ''' Gets a collection of theme objects.
    ''' </summary>
    Public Shared ReadOnly Property ThemesCatalog() As IEnumerable(Of ThemeCatalogItem)
        Get
            Return New ThemeCatalogItem() {BubbleCreme, TwilightBlue, ExpressionDark, ExpressionLight, WhistlerBlue, BureauBlack, BureauBlue, ShinyBlue, ShinyRed, RainierOrange, RainierPurple}
        End Get
    End Property

    ''' <summary>
    ''' Create a Theme background LinearGradientBrush.
    ''' </summary>
    ''' <param name="startColor">
    ''' Represents the start color of the gradient.
    ''' </param>
    ''' <param name="endColor">
    ''' Represents the end color of the gradient.
    ''' </param>
    ''' <returns>Returns a linear gradient brush.</returns>
    Private Shared Function GetLinearBrush(ByVal startColor As Color, ByVal endColor As Color) As LinearGradientBrush
        Dim brush As New LinearGradientBrush()
        Dim colorStop1 As New GradientStop()
        colorStop1.Color = startColor
        Dim colorStop2 As New GradientStop()
        colorStop2.Color = endColor
        colorStop2.Offset = 1
        brush.StartPoint = New System.Windows.Point(0.5, 1)
        brush.EndPoint = New System.Windows.Point(0.5, 0)
        brush.GradientStops = New GradientStopCollection()
        brush.GradientStops.Add(colorStop1)
        brush.GradientStops.Add(colorStop2)
        Return brush
    End Function
#End Region ' Sample Data
End Class
